# COGS 108 Group 171 Project

In [18]:
#Import the necessary libraries to begin project.
import requests
import json
import time
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Getting Metacritic Scores for Games

Here, we can start retrieving some Metacritic scores for some games from a dataset. This dataset can be found at data.world website, where the data is consisting of 13,357 rows. What we need is the game titles, and there corresponding Metacritic scores, along with the date they were released on.

In [19]:
metacritic_df = pd.read_csv('https://query.data.world/s/odkffdqc52dfm34xjwdc67q7cxfjmz')

In [20]:
metacritic_df = metacritic_df[metacritic_df['Metacritic'] != 0]
# Drop all columns besides the game name and the Metacritic score
metacritic_df = metacritic_df[['ResponseName', 'Metacritic', 'ReleaseDate']]
# This dataset has duplicates, so we need to drop those as well
metacritic_df = metacritic_df.drop_duplicates(subset=['ResponseName'])
# Rename ResonseName to GameTitle
metacritic_df.rename(columns={'ResponseName' : 'GameTitle'}, inplace=True)

In [21]:
metacritic_df.head()

GameTitle  Metacritic ReleaseDate
0                  Counter-Strike          88  Nov 1 2000
2                   Day of Defeat          79  May 1 2003
6                       Half-Life          96  Nov 8 1998
7  Counter-Strike: Condition Zero          65  Mar 1 2004
9           Half-Life: Blue Shift          71  Jun 1 2001

In [5]:
metacritic_df.shape

(2194, 3)

Here we can see that we have 2,194 game titles in our dataset with valid Metacritic scores.

## Retrieving data from Twitch

Now we need to retrieve data from the Twitch API. First, we need to get the `game_id` of each game, and then use those IDs to find the top 20 real-time streams per game.

In [22]:
# This is the data needed to authorize get requests from the Twitch API
data = {'Client-ID' : 'gaov4kdyvd87a5n7vnwn1d49hyv8lk'}

In [23]:
# Create a temporary array to store all the game titles from the metacritic_df dataframe
temp_arr = []
temp_arr = np.array(metacritic_df['GameTitle'])

# Create and initialize the main dataframe
df = pd.DataFrame()
    
# Loop through 100 games at a time, and for each iteration, append the API result to game_info
for i in range(0,temp_arr.size,100):
    
    query_names = temp_arr[i:i+100]
    
    # Send a GET request for 100 game titles
    r = requests.get('https://api.twitch.tv/helix/games', headers=data, params={'name': query_names})
    response = r.json()

    # Write the request data into specified file, and read it into a temporary dataframe
    with open('game_id_info.json','w') as outfile:
        json.dump(response, outfile)
        
    temp_df = pd.read_json('game_id_info.json', orient='index', typ='series')
    temp_df = temp_df['data']
    temp_df = pd.DataFrame(temp_df)
    
    # Append the temporary dataframe to the main dataframe   
    df = df.append(temp_df)



In [24]:
df

box_art_url      id  \
0   https://static-cdn.jtvnw.net/ttv-boxart/Space%...   11956   
1   https://static-cdn.jtvnw.net/ttv-boxart/Rag%20...    1275   
2   https://static-cdn.jtvnw.net/ttv-boxart/Left%2...   18916   
3   https://static-cdn.jtvnw.net/ttv-boxart/Multiw...   19711   
4   https://static-cdn.jtvnw.net/ttv-boxart/./Call...   11609   
5   https://static-cdn.jtvnw.net/ttv-boxart/./Safe...    7867   
6   https://static-cdn.jtvnw.net/ttv-boxart/Shadow...    8437   
7   https://static-cdn.jtvnw.net/ttv-boxart/Day%20...    5567   
8   https://static-cdn.jtvnw.net/ttv-boxart/./Cond...    8068   
9   https://static-cdn.jtvnw.net/ttv-boxart/X%20Re...   31315   
10  https://static-cdn.jtvnw.net/ttv-boxart/Half-L...    2755   
11  https://static-cdn.jtvnw.net/ttv-boxart/Arx%20...    9899   
12  https://static-cdn.jtvnw.net/ttv-boxart/Dota%2...   29595   
13  https://static-cdn.jtvnw.net/ttv-boxart/Dark%2...   13826   
14  https://static-cdn.jtvnw.net/ttv-boxart/./Vamp...   15062   
15  https://static-cdn.jtvnw.net/ttv-boxart/./X:%2...   11309   
16  https://static-cdn.jtvnw.net/ttv-boxart/./Full...    2242   
17  https://static-cdn.jtvnw.net/ttv-boxart/Team%2...   16676   
18  https://static-cdn.jtvnw.net/ttv-boxart/Half-L...    1420   
19  https://static-cdn.jtvnw.net/ttv-boxart/Killin...   23582   
20  https://static-cdn.jtvnw.net/ttv-boxart/./SiN%...    3914   
21  https://static-cdn.jtvnw.net/ttv-boxart/Portal...   19731   
22  https://static-cdn.jtvnw.net/ttv-boxart/Left%2...   24193   
23  https://static-cdn.jtvnw.net/ttv-boxart/Genesi...   13626   
24  https://static-cdn.jtvnw.net/ttv-boxart/Advent...    8523   
25  https://static-cdn.jtvnw.net/ttv-boxart/Psycho...    2673   
26  https://static-cdn.jtvnw.net/ttv-boxart/./S.T....   10775   
27  https://static-cdn.jtvnw.net/ttv-boxart/Bejewe...   22177   
28  https://static-cdn.jtvnw.net/ttv-boxart/BloodR...    1495   
29  https://static-cdn.jtvnw.net/ttv-boxart/./DEFC...    6109   
..                                                ...     ...   
42  https://static-cdn.jtvnw.net/ttv-boxart/The%20...  493044   
43  https://static-cdn.jtvnw.net/ttv-boxart/Dishon...  490348   
44  https://static-cdn.jtvnw.net/ttv-boxart/Starde...  490744   
45  https://static-cdn.jtvnw.net/ttv-boxart/Obscur...  492135   
46  https://static-cdn.jtvnw.net/ttv-boxart/Wailin...  494158   
47  https://static-cdn.jtvnw.net/ttv-boxart/The%20...  492765   
48  https://static-cdn.jtvnw.net/ttv-boxart/Trials...  492932   
49  https://static-cdn.jtvnw.net/ttv-boxart/The%20...  500711   
50  https://static-cdn.jtvnw.net/ttv-boxart/Shatte...  491920   
51  https://static-cdn.jtvnw.net/ttv-boxart/Steins...   25458   
52  https://static-cdn.jtvnw.net/ttv-boxart/Zombie...  492156   
53  https://static-cdn.jtvnw.net/ttv-boxart/Samoro...   31433   
54  https://static-cdn.jtvnw.net/ttv-boxart/March%...  492291   
55  https://static-cdn.jtvnw.net/ttv-boxart/IS%20D...  491465   
56  https://static-cdn.jtvnw.net/ttv-boxart/Pro%20...  493066   
57  https://static-cdn.jtvnw.net/ttv-boxart/Eagle%...  492401   
58  https://static-cdn.jtvnw.net/ttv-boxart/Spellw...  489189   
59  https://static-cdn.jtvnw.net/ttv-boxart/./Huma...  492504   
60  https://static-cdn.jtvnw.net/ttv-boxart/Bohemi...  496485   
61  https://static-cdn.jtvnw.net/ttv-boxart/Neon%2...  491951   
62  https://static-cdn.jtvnw.net/ttv-boxart/./Dang...   89297   
63  https://static-cdn.jtvnw.net/ttv-boxart/Senran...   68059   
64  https://static-cdn.jtvnw.net/ttv-boxart/RIVE-{...  493053   
65  https://static-cdn.jtvnw.net/ttv-boxart/OmniBu...  491699   
66  https://static-cdn.jtvnw.net/ttv-boxart/Shiny-...  492659   
67  https://static-cdn.jtvnw.net/ttv-boxart/Dofus-...   20596   
68  https://static-cdn.jtvnw.net/ttv-boxart/./Mega...  492270   
69  https://static-cdn.jtvnw.net/ttv-boxart/The%20...  491755   
70  https://static-cdn.jtvnw.net/ttv-boxart/Valent...  492615   
71  https://static-cdn.jtvnw.net/ttv-boxart/Deponi...  491671   

           

Great, we got over 1,560 games returned with their corresponding `game_id`s! Now, we can take each ID and return the top 20 streams of each game being broadcasted currently on Twitch.

In [26]:
# Reuse temp_arr to store each game ID
temp_arr = np.array(df['id'])

stream_df = pd.DataFrame()

# Counter for each request
count = 0

# Loop through 100 games at a time, and for each iteration, append the API result to game_info
for i in range(0,60):
    
    #query_ids = temp_arr[i:i+100]
    
    # Send a GET request for 100 game titles
    r = requests.get('https://api.twitch.tv/helix/streams', headers=data, params={'game_id': str(temp_arr[i])})
    response = r.json()

    # Write the request data into specified file, and read it into a temporary dataframe
    with open('game_stream_info.json','w') as outfile:
        json.dump(response, outfile)
        
    temp_df = pd.read_json('game_stream_info.json', orient='index', typ='series')
    temp_df = temp_df['data']
    temp_df = pd.DataFrame(temp_df)
    
    # Append the temporary dataframe to the main dataframe   
    stream_df = stream_df.append(temp_df)
    
    # Since the server limits us from obtaining more than 30 requests per minute, we need to 
    # pause the loop for one minute after 30 requests.
    if count is 30:
        time.sleep(60) # Wait one minute to obtain 30 more requests from Twitch servers
        count = 0
    else:
        count+=1


In [28]:
stream_df['game_id'].value_counts()
stream_df

community_ids game_id           id  \
0                                                  []   18916  34338895328   
1                                                  []   18916  34335575696   
2                                                  []   18916  34338909808   
0   [8deb7b0f-d5a3-4f9d-942d-2331d8f4fe3d, d808cf0...    2755  34334332432   
1   [d808cf0b-88c3-4ef2-acd0-bae82f17e9c3, dbec2d3...    2755  34337161872   
2              [50aab9c5-b677-4d9a-8b45-7c1c75dd6a99]    2755  34334823120   
3   [aca761bd-9eb9-4197-865f-2937995be149, b0e7cf1...    2755  34338312736   
0                                                  []    9899  34337037392   
0                                                  []   29595  34332372096   
1                                                  []   29595  34336712736   
2                                                  []   29595  34336295424   
3                                                  []   29595  34332824464   
4                                                  []   29595  34337173456   
5                                                  []   29595  34337999952   
6                                                  []   29595  34335904048   
7              [e071ba32-72ee-42d0-b488-c77febe8b919]   29595  34337193424   
8                                                  []   29595  34338268608   
9                                                  []   29595  34334788272   
10                                                 []   29595  34336031136   
11                                                 []   29595  34338775456   
12                                                 []   29595  34332380016   
13                                                 []   29595  34335971120   
14             [c0294643-085c-4943-94ab-f0e1ba090e23]   29595  34333678048   
15                                                 []   29595  34336890160   
16                                                 []   29595  34336701024   
17             [bfd58792-7d80-409c-9501-1daeeec58be8]   29595  34337115232   
18                                                 []   29595  34334333376   
19  [2785a0f0-2a67-4840-a1c3-7ad5e60a7d1f, e05ab43...   29595  34332503344   
0              [d808cf0b-88c3-4ef2-acd0-bae82f17e9c3]   13826  34338804736   
1                                                  []   13826  34338127984   
..                                                ...     ...          ...   
9                                                  []   10775  34338835264   
0                                                  []   17343  34338967280   
0                                                  []   13158  34337031088   
0                                                  []    5250  34338294432   
1                                                  []    5250  34337460048   
0                                                  []   15067  34338675744   
0   [2f27ff6d-97fb-4df7-886d-c7893a13a13f, 4bef9cc...   32399  34335702464   
1   [2f27ff6d-97fb-4df7-886d-c7893a13a13f, 4bef9cc...   32399  34335709920   
2   [2f27ff6d-97fb-4df7-886d-c7893a13a13f, 4bef9cc...   32399  34335913904   
3                                                  []   32399  34336653744   
4                                                  []   32399  34337132736   
5                                                  []   32399  34338780400   
6              [84f777d5-c349-41d6-bbda-3605d1f3e46b]   32399  34332816448   
7   [3eac753c-2e37-4970-9145-c1767607f044, 4bef9cc...   32399  34338685504   
8                                                  []   32399  34334343728   
9                                                  []   32399  34336058064   
10             [2f27ff6d-97fb-4df7-886d-c7893a13a13f]   32399  34336063600   
11             [db31eae1-8f75-44a9-af19-1ec275163945]   32399  34335097808   
12                                                 []   32399  34336027984   
13             [4bef9cc9-c81d-4384-a62e-b16cf5e7a45e]   32399 

### Test code API

Here is a test example of retrieving data from the Twitch Web API. Let's see if we can extract a JSON file from a URL.

Let's try to retrieve the data and import it into a pandas dataframe.

In [ ]:
# Send a GET request to the API to get all the streams for GTA V
r = requests.get('https://api.twitch.tv/helix/streams', headers=data, params={'game_id': ['16676', '29595']})
# load the data as a json into the variable 'response'
response = r.json()
# write the data to the file 'twitch_data.json'
with open('twitch_data.json','w') as outfile:
    json.dump(response, outfile)

In [ ]:
df_test = pd.read_json('twitch_data.json', orient='index', typ='series')

In [ ]:
df_test = df_test['data']
df_test = pd.DataFrame(df_test)
df_test

In [ ]:
df_test.shape